### PR TITLE
Best practices for keeping enclaves secure

### PR DESCRIPTION
**Pull request by douglasmaciver**
# Best Practices for Secure Enclaves
Use this guide to apply secure patterns and avoid common mistakes that put the security of your enclave applications at risk.

Credit to Ming-Wei for early code samples and text. @mingweishih

### CODE REVIEW

**Comment by ghost** [comment]
Welcome @douglasmaciver! <br><br>It looks like this is your first PR to <a href='https://github.com/openenclave/openenclave-security'>openenclave/openenclave-security</a> 🎉🎉. Please refer to our [pull request process documentation](https://github.com/openenclave/test-infra/blob/master/docs/owners.md#the-code-review-process) to help your PR have a smooth ride to approval. <br><br>You will be prompted by a bot to use commands during the review process. Do not be afraid to follow the prompts! It is okay to experiment. [Here is the bot commands documentation](https://go.k8s.io/bot-commands). <br><br>You can also check if openenclave/openenclave-security has [its own contribution guidelines](https://github.com/openenclave/openenclave-security/tree/master/CONTRIBUTING.md). <br><br>Thank you, and welcome to openenclave. :smiley:

**Comment by ghost** [comment]
@douglasmaciver: GitHub didn't allow me to assign the following users: swamymsft.

Note that only [openenclave members](https://github.com/orgs/openenclave/people), repo collaborators and people who have commented on this issue/PR can be assigned. Additionally, issues/PRs can only have 10 assignees at the same time.
For more information please see [the contributor guide](https://git.k8s.io/community/contributors/guide/first-contribution.md#issue-assignment-in-github)

<details>

In response to [this](https://github.com/openenclave/openenclave-security/pull/9#issuecomment-879476920):

>/assign @swamymsft


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (swamymsft) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Enclaves are already considered as secure, should we call this something like "Security best practices to follow during enclave development"

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I am working on this now.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: TODO: GH relative path needs to be updated.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: TODO: extraneous ) at end of line.

**Comment by ghost** [comment]
@douglasmaciver: GitHub didn't allow me to assign the following users: swamymsft.

Note that only [openenclave members](https://github.com/orgs/openenclave/people), repo collaborators and people who have commented on this issue/PR can be assigned. Additionally, issues/PRs can only have 10 assignees at the same time.
For more information please see [the contributor guide](https://git.k8s.io/community/contributors/guide/first-contribution.md#issue-assignment-in-github)

<details>

In response to [this](https://github.com/openenclave/openenclave-security/pull/9#issuecomment-880019874):

>/assign @swamymsft


Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: How about "Best practices for keeping enclaves secure."?

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I don't think this application of volatile is correct. Furthermore, we have to consider which behavioral form (ISO, MS) is being used. 
I'm looking into this now.

Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:125]: volatile belongs on blob, not captured_blob. that's the memory that could change at anytime.

Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I think the current approach is okay, but I'm going change the approach a bit: instead of using the blob struct, I'm going switch to two volatile vars.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (anakrish) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: > Marshaling parameters from the host to the enclave for security purposed, and in doing so also helps to mitigate certain processor vulnerabilities (like spectre).

I'm not able to easily parse this sentence.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (anakrish) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Can't we just do `captured_blob = *blob` ?

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (anakrish) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:148]: Alternatively we can make the host object pointer volatile: `volatile blob_t* blob`.
The compiler cannot introduce extra reads of the volatile variable.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (anakrish) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]: Could be worth clarifying that Open Enclave does not provide code confidentiality.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:148]: Thanks Anand, I decided to follow the precedent in `oe_ecall_context_get_ocall_buffer`().

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]: Good point Anand. I added:

> It's important to note that enclaves do not provide _code confidentiality_. This is noteworthy in the context of this discussion because, as you will read, hard-coded credentials in application code are to be avoided -- with or without enclave protections.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Yes. That would have been better. But now it's moot for we have adopted the `oe_ecall_context_get_ocall_buffer()` precedent.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: I agree, that sentence and paragraph need some reworking:

> Calling into and out of enclaves uses special methods that orchestrate the context switch and marshal the function parameters. Much of the code necessary to manage these calls and parameter marshaling are common to most applications. The Open Enclave _oeedger8r_, using the interface definition defined by application developers in EDL files, generates boilerplate code for them.

**Comment by douglasmaciver** [comment]
@swamymsft @mrragava @anakrish Thank you for your review comments. I acted on most of the comments.

I just checked-in my latest updates.  They address comments made so far.

At this point, the material has had enough initial review and we should proceed to a PR for openenclave.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (swamymsft) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Please assign the numbering to match the list above

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (swamymsft) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:88]: You might want to consider mentioning in the comments above about this library routine

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (swamymsft) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:94]: Significance of comparing the size to 4096 bytes is not obvious

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (swamymsft) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: This can be build as table to include the sub items that are listed under each of the best practices

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (swamymsft) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]: what about OP-TEE?

**Review by swamymsft — APPROVED** [review]

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:88]: Good idea: I added:
> The Open Enclave SDK provides two library functions for checking memory buffers: `oe_is_within_enclave()` and `oe_is_outside_enclave()`.

> Let's update the code to ensure that the memory block is at least the size of a `blob_t` and is located _strictly_ outside the secured enclave memory region before operating on it by calling `oe_is_outside_enclave()`.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Done.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Done

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:94]: I added a comment.
>    // Blobs to be rendered are limited to 4K.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]: I changed the oe_is_within_enclave() != !oe_is_outside_enclave() section to be specific to SGX: 
> 1.2 SGX-based enclaves: oe_is_within_enclave() != !oe_is_outside_enclave()

@swamymsft : Are there other concerns about OP-TEE?

**Review by mrragava — APPROVED** [review]

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [inline docs/GettingStartedDocs/best_practices_secure_enclaves.md]: There's an incompatibility between Visual Code and Github's markdown:  Github's parser does not like something in the link. I will have to experiment more tomorrow.

**Comment by mrragava** [comment]
Could you rebase and squash the commits.

**Comment by douglasmaciver** [comment]
@mrragava I will rebase and squash tomorrow after I fix the minor link problem.

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md]: Finally!

**Review by swamymsft — APPROVED** [review]

File: docs/GettingStartedDocs/best_practices_secure_enclaves.md
Comment (douglasmaciver) [reply docs/GettingStartedDocs/best_practices_secure_enclaves.md:246]: We discussed this during our daily stand-up. Adjusting the focus of the boundary calls to SGX is sufficient.

**Comment by douglasmaciver** [comment]
I cleaned up the PR...it's now a single commit.

**Review by mrragava — APPROVED** [review]

**Comment by ghost** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/openenclave/openenclave-security/pull/9#" title="Author self-approved">douglasmaciver</a>*, *<a href="https://github.com/openenclave/openenclave-security/pull/9#pullrequestreview-710122669" title="Approved">mrragava</a>*, *<a href="https://github.com/openenclave/openenclave-security/pull/9#pullrequestreview-709876526" title="Approved">swamymsft</a>*

The full list of commands accepted by this bot can be found [here](?repo=openenclave%2Fopenenclave-security).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/openenclave/openenclave-security/blob/main/OWNERS)~~ [mrragava]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
